File: elixir/test/moon_direct_test.exs
# Moon direct test (offline, mock system.fetch)

defmodule Solardemo.MoonDirectTest do
  use ExUnit.Case

  alias Voxgig.Struct, as: S
  alias Solardemo.Helpers, as: H

  # A recording live-mode SDK whose transport returns `canned`.
  defp mk(canned) do
    parent = self()

    fetch = fn url, _fetchdef ->
      send(parent, {:called, url})
      {S.jm(["status", 200, "statusText", "OK", "headers", S.jm([]), "json", fn -> canned end, "body", "mock"]), nil}
    end

    Solardemo.new(S.jm(["base", "http://localhost:8080", "system", S.jm(["fetch", fetch])]))
  end

  test "should direct list moon" do
    canned = S.jt([S.jm(["id", "direct01"]), S.jm(["id", "direct02"])])
    sdk = mk(canned)
    res = Solardemo.direct(sdk, H.deep(%{"path" => "/moon", "method" => "GET", "params" => %{}}))
    assert S.getprop(res, "ok") == true
    assert H.to_int(S.getprop(res, "status")) == 200
    assert S.islist(S.getprop(res, "data"))
    assert S.size(S.getprop(res, "data")) == 2
    assert_received {:called, _url}
  end

  test "should direct load moon" do
    canned = S.jm(["id", "direct01"])
    sdk = mk(canned)
    res = Solardemo.direct(sdk, H.deep(%{"path" => "/moon/direct01", "method" => "GET", "params" => %{}}))
    assert S.getprop(res, "ok") == true
    assert H.to_int(S.getprop(res, "status")) == 200
    data = S.getprop(res, "data")
    assert S.ismap(data)
    assert S.getprop(data, "id") == "direct01"
    assert_received {:called, _url}
  end
end

File: ocaml/test/moon_direct_test.ml
(* Generated moon direct-call test. *)

open Voxgig_struct
open Sdk_types
open Sdk_helpers
open Testutil

let () =
  test "moon.direct_call" (fun () ->
      let calls = ref [] in
      let mock = Func (fun _ args _ _ ->
          calls := !calls @ [args];
          jo [("status", Num 200.); ("statusText", Str "OK"); ("headers", empty_map ());
              ("json", json_thunk (jo [("id", Str "direct01")]))]) in
      let client = Sdk_client.make
          (jo [("base", Str "http://localhost:8080"); ("system", jo [("fetch", mock)])]) in
      let result = Sdk_client.direct client
          (jo [("path", Str "/moon"); ("method", Str "GET")]) in
      check_vbool "ok" (getp result "ok") true;
      check_vstr "data.id" (getp (getp result "data") "id") "direct01";
      check_int "one call" (List.length !calls) 1)
